### PR TITLE
Refactor redux store to support multiple plans

### PR DESCRIPTION
**Pull request by NilsBacke**
Overview of changes:
Remove scheduleReducer and scheduleActions
Replace schedulesReducer and schedulesActions with userPlansReducer and userPlansActions
Rename all state getters to be more explicit and end in "FromState"
Rename all actions to be more explicit and end in "Action"

Add completed courses, transfer courses, and exam credits to the user store (as they are specific to a user, not a plan)

### CODE REVIEW

File: common/types.ts [hunk 1]
```diff
@@ -260,7 +260,6 @@ export interface Schedule {
   yearMap: {
     [key: number]: ScheduleYear;
   };
-  id: string;
```
Comment (NilsBacke) [inline common/types.ts:263]: We don't need the id field on a schedule anymore

File: frontend/src/Onboarding/LoginScreen.ts [hunk 2]
```diff
@@ -0,0 +1,350 @@
+export {};
```
Comment (NilsBacke) [inline frontend/src/Onboarding/LoginScreen.ts:1]: Going to comment this whole file out for now, as we won't have to worry about it (until we expand to other colleges)

File: frontend/src/Onboarding/Onboarding.tsx [hunk 3]
```diff
@@ -255,14 +251,3 @@ class OnboardingComponent extends React.Component<Props> {
     );
   }
 }
-
-const mapStateToProps = (state: AppState) => ({
-  fullName: getFullNameFromState(state),
-});
```
Comment (NilsBacke) [inline frontend/src/Onboarding/Onboarding.tsx:261]: Unused variables

File: frontend/src/Onboarding/OnboardingInfoScreen.tsx [hunk 4]
```diff
@@ -337,21 +330,21 @@ const mapDispatchToProps = (dispatch: Dispatch) => ({
     dispatch(setAcademicYearAction(academicYear)),
   setGraduationYear: (academicYear: number) =>
     dispatch(setGraduationYearAction(academicYear)),
-  setMajor: (major?: Major) => dispatch(setDeclaredMajorAction(major)),
-  setCoopCycle: (coopCycle: string, plan: Schedule) =>
-    dispatch(setCoopCycle(coopCycle, plan)),
+  setMajor: (major?: Major) => dispatch(setUserMajorAction(major?.name || '')),
```
Comment (NilsBacke) [inline frontend/src/Onboarding/OnboardingInfoScreen.tsx:333]: We only store the major name in redux (to stay consistent with the backend), and we can always use `findMajorFromName` if we need the object

File: frontend/src/Onboarding/RedirectScreen.tsx [hunk 5]
```diff
@@ -90,17 +80,7 @@ export const RedirectScreen: React.FC<Props> = ({ redirectUrl }) => {
...
-          dispatch(setAcademicYearAction(response.user.academicYear));
-          const maj = findMajorFromName(response.user.major, majors);
-          if (maj) {
-            dispatch(setDeclaredMajorAction(maj));
-          }
-          dispatch(setUserIdAction(response.user.id));
-          dispatch(setEmailAction(response.user.email));
-          dispatch(setUserCoopCycleAction(response.user.coopCycle));
-          dispatch(setIsAdvisorAction(response.user.isAdvisor));
+          dispatch(setUserAction(response.user));
```
Comment (NilsBacke) [inline frontend/src/Onboarding/RedirectScreen.tsx:83]: We can just set the whole user object instead of setting each field individually

Comment (NilsBacke) [reply frontend/src/Onboarding/RedirectScreen.tsx:83]: Keep in mind this requires that the fields in the user object in redux must match the user object returned from the backend

File: frontend/src/components/TransferCreditHolder.tsx [hunk 6]
```diff
@@ -190,7 +185,6 @@ const mapDispatchToProps = (dispatch: Dispatch) => ({
     dispatch(addTransferClassAction(courses)),
   onDeleteClass: (course: ScheduleCourse) =>
     dispatch(removeTransferClassAction(course)),
-  onUndoDeleteClass: () => dispatch(undoRemoveClassAction()),
```
Comment (NilsBacke) [inline frontend/src/components/TransferCreditHolder.tsx:193]: I created another trello card to re-implement undoing deleting a transfer class

File: frontend/src/home/Home.tsx [hunk 7]
```diff
@@ -275,32 +243,17 @@ class HomeComponent extends React.Component<Props, HomeState> {
...
-          });
-        }
-      );
-    } else if (this.props.activeSchedule) {
-      this.setState({ fetchedPlan: true });
-    }
+    const token = getAuthToken();
+    findAllPlansForUser(this.props.userId!, token).then(
+      (plans: IPlanData[]) => {
+        this.props.setUserPlans(plans, this.props.academicYear);
```
Comment (NilsBacke) [inline frontend/src/home/Home.tsx:249]: This also sets the active plan

File: frontend/src/home/Home.tsx [hunk 8]
```diff
@@ -313,23 +266,25 @@ class HomeComponent extends React.Component<Props, HomeState> {
...
-      );
-      if (!this.props.closedYears.has(yearIdx)) {
-        numVisibleWarnings++;
-        if (numVisibleWarnings <= 5) {
-          // add new toasts
-          this.props.addToast(w.message, {
-            appearance: "warning",
-            autoDismiss: true,
-          });
+    if (this.props.activePlan) {
```
Comment (NilsBacke) [inline frontend/src/home/Home.tsx:269]: Only do this is we have an active plan. There is a time while the user plans are loading in the home screen that there is not an active plan, so we only want to render things that do not rely on an active plan while the user plans are loading

File: frontend/src/home/Home.tsx [hunk 9]
```diff
@@ -338,20 +293,20 @@ class HomeComponent extends React.Component<Props, HomeState> {
     // if drag is coming from the sidebar
     if (isNaN(Number(source.droppableId))) {
       addCourseFromSidebar(
-        this.props.schedule,
+        this.props.activePlan!.schedule,
```
Comment (NilsBacke) [inline frontend/src/home/Home.tsx:296]: In order to be able to add a course from sidebar, there must be an active plan, so it's okay to assert here.

File: frontend/src/home/Home.tsx [hunk 10]
```diff
@@ -476,82 +437,25 @@ class HomeComponent extends React.Component<Props, HomeState> {
...
       }
     });
   }
 
-  /**
-   * If a user is currently logged in, saves the current plan under this user.
-   * Only supports updating a user's singular plan, can be modified later to
-   * update a specific plan.
-   */
-  savePlan() {
```
Comment (NilsBacke) [inline frontend/src/home/Home.tsx:524]: Remove this unused function

File: frontend/src/home/Home.tsx [hunk 11]
```diff
@@ -564,6 +468,20 @@ class HomeComponent extends React.Component<Props, HomeState> {
   };
 
   render() {
+    if (!this.state.fetchedPlan) {
```
Comment (NilsBacke) [inline frontend/src/home/Home.tsx:471]: To air on the safe side, just render a loading indicator if currently fetching user plans

File: frontend/src/profile/Profile.tsx [hunk 12]
```diff
@@ -140,23 +136,23 @@ interface ChangePasswordProps {
...
-          value={props.name}
-          onChange={e => props.setName(e.target.value)}
-          placeholder="John Smith"
-        />
-      )}
-      {!props.isEdit && <ItemEntry> {props.name} </ItemEntry>}
-    </ProfileEntryContainer>
-  );
-};
+// const ProfileName = (props: ProfileNameProps) => {
```
Comment (NilsBacke) [inline frontend/src/profile/Profile.tsx:139]: Going to re-implement this in a separate PR, as right now we don't even show the Profile page. Right now we want to show their full name, but not allow them to edit it.

**Comment by NilsBacke** [comment]
[Get ScheduleAction/Reducer renamed so code is readable and clean up reducers (a lot)](https://trello.com/c/LuJKx21c/151-get-scheduleaction-reducer-renamed-so-code-is-readable-and-clean-up-reducers-a-lot)

File: frontend/src/Onboarding/RedirectScreen.tsx (hunk 5 above)
Comment (ArunJ98) [reply frontend/src/Onboarding/RedirectScreen.tsx:83]: don't think this will work for transfer/completed courses. What happens if we pass the response and it doesn't match, will it just not set those fields or will it error?

File: frontend/src/profile/Profile.tsx (hunk 12 above)
Comment (ArunJ98) [reply frontend/src/profile/Profile.tsx:139]: yea we might want to rethink the UI/purpose of this page, probably not super high priority right now

File: frontend/src/home/Home.tsx (hunk 7 above)
Comment (ArunJ98) [reply frontend/src/home/Home.tsx:249]: how do we determine which one is the active plan?

File: frontend/src/home/Home.tsx (hunk 8 above)
Comment (ArunJ98) [reply frontend/src/home/Home.tsx:269]: nice

File: frontend/src/state/index.ts [hunk 13]
```diff
@@ -176,66 +137,87 @@ export const getTakenCredits = (state: AppState): number =>
...
-export const getIsAdvisorFromState = (state: AppState): boolean | undefined => {
-  return state.user.isAdvisor;
+export const getIsAdvisorFromState = (state: AppState) => {
+  return getUserFromState(state).isAdvisor;
 };
+
+/*
+ * SAFE GET FUNCTIONS
+ * Should only need to use these where there is not guaranteed to have an active plan (like the home screen)
+ * TODO: look into if a separate component (prior to getting to the HomeComponent) to load plan data would be worth it
```
Comment (ArunJ98) [inline frontend/src/state/index.ts:209]: What if we just loaded all plan info on the loading screen on the redirect page. Once we get the user, we can get all necessary plan info/etc... We can do this in a separate PR but I think this would simplify the flow.

Comment (NilsBacke) [reply frontend/src/state/index.ts:209]: I did consider this when coming up with solutions, it seemed like a lot of changes at the time. I do think that your idea is the best though, basically for the reason that you shouldn't be able to go to the HomeScreen without an active plan (which logically makes sense).

Comment (NilsBacke) [reply frontend/src/state/index.ts:209]: I added a card for this in Trello

**Comment by NilsBacke** [comment]
We determine which plan is active by just choosing the first plan for now (we do this in the reducer for setUserPlansAction). We have a card in trello to add a `lastViewed` timestamp field on a plan, and the most recently viewed plan will be made the active plan.

**Comment by NilsBacke** [comment]
"don't think this will work for transfer/completed courses. What happens if we pass the response and it doesn't match, will it just not set those fields or will it error?"

It won't error, it just won't save correctly. 

For completed/transfer courses could we just make that part of the PR you're working on? Right now this is the same functionality as in master right now, where we don't handle completed/transfer courses from the backend.

**Review by ArunJ98 — APPROVED** [review]

**Comment by NilsBacke** [comment]
[Fully support multiple plans in the frontend/backend](https://trello.com/c/xTtClswf/213-fully-support-multiple-plans-in-the-frontend-backend)
